nasa-veda-platform/scientific-computing/custom-environments.qmd

```diff
@@ -0,0 +1,171 @@
+- A **list of dependencies** to install, specified via various files (such as
+  `environment.yml`, `requirements.txt`, or `apt.txt`, among others), and are
+  installed **in addition to** a [base set of dependencies].
+- *Optionally*, a **post-build script** to run after all dependencies are
+  installed (this must be named [postBuild])
+- *Optionally*, a **pre-session script** to run each time you start a server
+  with your image (this must be named [start])
+
+Such files from your repository are used to construct a `Dockerfile` file on
+your behalf, so you don't have to write one yourself, unless your really want
+to.  If your repo contains a `Dockerfile`, repo2docker will ignore all other
+files in your repo that it would otherwise use to generate a `Dockerfile`, and
+simply use the `Dockerfile` from your repo.
+
+For complete details on what repo2docker supports, see the following:
+
+- [Where to put configuration files]
+- [Configuring your repository]
+- [Architecture]
+- [Frequently Asked Questions (FAQ)]
+- [Example repositories]
+
+Once you have described your image in your repository, you are ready to build
+it, as described in the next section.
+
+### Building Your Image
+
+When you choose **Build your own image** from the **Environment** list, you must
+specify values for 2 other fields --- **Repository** and **Git Ref** --- which
+indicate which repository to use and which commit within the repository to use
+(valid values for these fields are explained below the screenshot):
+
+![Specifying a repository for building your own image](images/environment-build-your-own-image.png)
+
+In the **Repository** field, you must specify a public GitHub repository in
+either of the following forms:
+
+- **Full URL** (e.g., <https://github.com/binder-examples/conda>)
+- **Namespace/name pair** (e.g., `binder-examples/conda`), where a namespace
+  is either a GitHub *username* or *organization* (`binder-examples` in this
+  example).
+
+In the **Git Ref** field, if you want repo2docker to use the most recent
+versions of files from the default branch of the repository (typically `main`),
+then use the default value **HEAD** to indicate the most recent commit.
+Alternatively, you may specify a **branch name**, a **tag**, or a specific
+**commit identifier** to indicate a different commit for repo2docker to use.
+
+Once you've specified both **Repository** and **Git Ref**, click the **Build
+image** button to trigger repo2docker to build your image (which may take
+several minutes), and you should see log messages appear below the **Build
+image** button, similar to the following:
+
+![Building a new image](images/build-logs-new-build-started.png)
+
+Once your image is built, you should see something similar to this:
+
+![Image build has finished](images/build-logs-new-build-finished.png)
+
+The next time you click **Build image** for the same values of **Repository**
+and **Git Ref**, if the Git Ref refers to the same commit in the repository as
```

Comment:
should we warn people against using 'latest' because it won't know the image has changed?

Reply:
Good idea. I'll add a warning about that.

Reply:
@wildintellect, I think you may be referring to the earlier section about using pre-build images vs. building a custom one in this section. Is that correct?

Either way, can you clarify?

If you're referring to the previous section, then I suspect that using "latest" actually would detect an image change, so that if the "latest" image changed since a previous environment launch, then a subsequent environment launch would use the newer image -- similar to how using "HEAD" here for Git Ref would cause a new custom image to be built if a new commit were pushed since the last build.

Reply:
It does check the repo for updates when you specify a ref that is not a specific commit. So using "HEAD" or a branch name will cause an image rebuild if the default (or specified) branch contains new commits since the last time the image was built. That's what this bit of the doc describes, although after re-reading, it could probably use a bit of rewording for better clarity.